client/rest-high-level/src/test/java/org/elasticsearch/client/RequestTests.java

```diff
@@ -155,4 +165,106 @@ private static void getAndExistsTest(Function<GetRequest, Request> requestConver
+        if (id != null) {
+            method = "PUT";
+            if (randomBoolean()) {
+                indexRequest.opType(DocWriteRequest.OpType.CREATE);
+                if (version ==  Versions.MATCH_ANY) {
+                    // When op_type is create and version match_any, version is deleted (see IndexRequest.version())
+                    expectedParams.put("version", Long.toString(Versions.MATCH_DELETED));
+                }
+            }
+        }
+
+        if (randomBoolean()) {
+            String timeout = randomTimeValue();
+            indexRequest.timeout(timeout);
+            expectedParams.put("timeout", timeout);
+        } else {
+            expectedParams.put("timeout", ReplicationRequest.DEFAULT_TIMEOUT.getStringRep());
+        }
+
+        if (frequently()) {
+            if (randomBoolean()) {
+                String routing = randomAsciiOfLengthBetween(3, 10);
+                indexRequest.routing(routing);
+                expectedParams.put("routing", routing);
+            }
+            if (randomBoolean()) {
+                String parent = randomAsciiOfLengthBetween(3, 10);
+                indexRequest.parent(parent);
+                expectedParams.put("parent", parent);
+            }
+            if (randomBoolean()) {
+                String pipeline = randomAsciiOfLengthBetween(3, 10);
+                indexRequest.setPipeline(pipeline);
+                expectedParams.put("pipeline", pipeline);
+            }
+
+            if (randomBoolean()) {
+                WriteRequest.RefreshPolicy refreshPolicy = randomFrom(WriteRequest.RefreshPolicy.values());
+                indexRequest.setRefreshPolicy(refreshPolicy);
+                if (refreshPolicy != WriteRequest.RefreshPolicy.NONE) {
+                    // TODO use refreshPolicy.toString() once it returns the appropriate value
+                    if (refreshPolicy == WriteRequest.RefreshPolicy.IMMEDIATE) {
+                        expectedParams.put("refresh", Boolean.TRUE.toString());
+                    } else {
+                        expectedParams.put("refresh", "wait_for");
+                    }
+                }
+            }
+            if (randomBoolean()) {
+                VersionType versionType = randomFrom(VersionType.values());
+                indexRequest.versionType(versionType);
+                if (versionType != VersionType.INTERNAL) {
+                    expectedParams.put("version_type", versionType.name().toLowerCase(Locale.ROOT));
+                }
+            }
+        }
+
+        XContentType xContentType = randomFrom(XContentType.values());
+        int nbFields = randomIntBetween(0, 10);
+        try (XContentBuilder builder = XContentBuilder.builder(xContentType.xContent())) {
+            builder.startObject();
+            for (int i = 0; i < nbFields; i++) {
+                builder.field("field_" + i, i);
+            }
+            builder.endObject();
+            indexRequest.source(builder);
```

Comment:
I don't know if it is compatible, but maybe you can use RandomObjects#randomSource(Random random) here instead?